### PR TITLE
[notebook2]: notebook state management

### PR DESCRIPTION
**Pull request by akotlar**
Stacks on #5452. See next comment for some additional details. 

This gets notebook2 into a state equivalent to that of notebook 1 for user-facing content, but with the styling of app.hail.is. 2 PRs remain: bring fine-grained state updates, remove services.

PR-specific commits to preview diff after #5452 merged:
https://github.com/hail-is/hail/pull/5476/commits/2f180ed0bfb3b0dfb7224df1ef6afba0e1a9cbfc
https://github.com/hail-is/hail/pull/5476/commits/a60780e506e77cddf6afbcf388f9bfb027a32f8f
https://github.com/hail-is/hail/pull/5476/commits/1d58947696043f16ad864ee38c40e6dccf87cb1a

edit notes: Added relevant commits.

### CODE REVIEW

File: notebook2/notebook/notebook.py [hunk 1]
```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
     return url + path
 
 
+# FIXME: use hash
+def UNSAFE_user_id_transform(user_id): return user_id.replace('|', '--_--')
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: Issue: https://github.com/hail-is/hail/issues/5377

File: notebook2/notebook/notebook.py [hunk 2]
```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
...
+
+
+def read_svc_status(svc_name: str):
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: This is a placeholder. Next PR will drop services

File: notebook2/notebook/notebook.py [hunk 3]
```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
...
+    try:
+        # TODO: inspect exception for non-404
+        _ = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: I propose improvements to this be a separate PR. This PR is to get the MVP up, there are surely nicer interfaces to develop.

File: notebook2/notebook/notebook.py [hunk 4]
```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
...
+        tm = {"exit_code": state.terminated.exit_code, "finished_at": state.terminated.finished_at,
+              "started_at": state.terminated.started_at, "reason": state.terminated.reason}
+
+    if rn is None and wt is None and tm is None:
+        return None
+
+    return {"running": rn, "terminated": tm, "waiting": wt}
+
+
+def read_conditions(conds):
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: I think it's nice for the user to have some insight into the container status. I believe, from our dev forum discussion pods remain running even after all containers terminate, but this was also brought up at our last-thursday conversation with @cseed, so thought I'd follow up in an issue to better document Kubernetes behavior. I've opened an issue for this: https://github.com/hail-is/hail/issues/5486

File: notebook2/notebook/notebook.py [hunk 5]
```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
...
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
+
+
+def marshall_notebooks(pods = [], svc_status = None):
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: Returns array because 1) in the future we may want to allow this 2) a deleted pod will not be cleared immediately, and we may either want the user to see this, somewhere (maybe behind a log dialog/slide-out element), or ignore such pods.

File: notebook2/notebook/notebook.py [hunk 6]
```diff
@@ -170,61 +198,201 @@ def external_url_for(path):
...
+        return render_template('notebook.html',
+                               form_action_url=flask.url_for('notebook_post'),
                                images=list(WORKER_IMAGES),
                                default='hail')
-    svc_name = session['svc_name']
-    jupyter_token = session['jupyter_token']
-    log.info('redirecting to ' + external_url_for(f'instance/{svc_name}/?token={jupyter_token}'))
-    return redirect(external_url_for(f'instance/{svc_name}/?token={jupyter_token}'))
 
+    # TODO: We currently simplify state tracking, since using vanilla js makes this trickier
```
Comment (akotlar) [inline notebook2/notebook/notebook.py]: Issue: https://github.com/hail-is/hail/issues/5487

**Comment by akotlar** [comment]
@danking 

This should be ready to look at. Stacks on #5452. Once that PR is merged, changes are:

1) notebook.html: organize into form (notebook-form.html) and notebook state (notebook-state.html) components.
2) add modified versions of notebook-api. Namely I don't use the marshaling procedure you didn't like, and refactor as much of the JS stuff as I can into synchronous http requests.
   * modifies /notebook routes , adds `marshall_notebook`, `get_live_user_notebooks`, `wait_websocket`, and replace any calls to `session['pod_name']` and `session['svc_name`] with equivalent versions based on `session['notebook']`, which contains the notebook object of the existing session

Basically feels like a synchronous / refresh-based version of what we had on app.hail.is, with less state insight (uses only the websocket-based reachability check).

Upcoming PR will restore fine-grained state updates via JS/websocket.

cc @cseed

Images:
screen shot 2019-02-28 at 3 06 46 pm
screen shot 2019-02-28 at 3 06 08 pm
screen shot 2019-02-28 at 3 06 51 pm

File: notebook2/notebook/notebook.py (hunk 2 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Also, this will disappear in the next PR (remove services)

File: notebook2/notebook/notebook.py (hunk 1 above)
Comment (danking) [reply notebook2/notebook/notebook.py]: I'm inclined to not leave a FIXME in since the method is already labeled UNSAFE and we have an issue.

Comment (akotlar) [reply notebook2/notebook/notebook.py]: Got it, agreed

Comment (akotlar) [reply notebook2/notebook/notebook.py]: Ok, so the reason I didn't do this yet...I'm a bit worried about the probability of collisions.

The reason I think this is unsafe is kind of for the same reason, that I will have collisions by replacing all "|" with some other pattern. 

Auth0 ids come in the form of provider|id. One solution is to replace only the first |. A better solution seems to encode into a form that kube will accept.

Do you have any recommendations?

**Comment by akotlar** [comment]
@danking addressed the unsafe issue, sha224 (largest that results in a <63 character hex digest)

File: notebook2/notebook/notebook.py (hunk 1 above)
Comment (danking) [reply notebook2/notebook/notebook.py]: heh, I really hate when people don't precisely define the syntax of their data.
Maybe this?
```python
x = re.match(r"^([^|]+)\|(.+)$", user_id)
identity_provider = x.group(1)
user_id = x.group(2)
identity_provider + '.' + user_id
```

? Is there an email or forum we can ask about the syntax of identity_providers? I've only seen dashes and characters.

Comment (akotlar) [reply notebook2/notebook/notebook.py]: Agreed, good suggestion. Did a bit of research.
1) Auth0's documentation states <provier>|<id> , but doesn't specify exactly what the id format is (as far as I can see): https://auth0.com/docs/tokens/id-token

2) [The OpenID Connect standard](https://openid.net/specs/openid-connect-core-1_0.html#StandardClaims), which Auth0 implements (although I'm not sure all of their APIs are compliant), specifies ASCII and 255 characters max, which would be disallowed by Kube's label restrictions, if full range is used.

3) [I opened an issue to ask](https://community.auth0.com/t/sub-claim-format-kubernetes-restricted-set/21996)

For now, I hashed using SHA224 and store the hex digest. What do you think of that?

Comment (danking) [reply notebook2/notebook/notebook.py]: Heh. The response side-stepped defining what the identity provider's syntax is. 🤷‍♀️ 

SHA seems fine.

Comment (danking) [reply notebook2/notebook/notebook.py]: Actually wait, why are we introducing a possibly colliding hash? We can just base64 encode the bytes, no?

```
In [3]: base64.b64decode(base64.b64encode("google-foogle|bobadob"))
Out[3]: 'google-foogle|bobadob'
```

We pay a 4/3x inflation, which leaves us about 189 characters for the tokens.

File: notebook2/notebook/templates/notebook-state.html [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+            close
+        </button>
+    </form>
+</div>
+{% if not ready %}
+    <script>
+        let protocol = location.protocol.replace("http", "ws")
+        let sock = new WebSocket(protocol + location.host + "/wait");
+        sock.onmessage = function (event) {
+            window.location.reload()
```
Comment (danking) [inline notebook2/notebook/templates/notebook-state.html]: Why reload this page? We receive an event iff the notebook is ready, we should should redirect to the notebook.

File: notebook2/notebook/notebook.py [hunk 8]
```diff
@@ -357,11 +492,13 @@ def worker_image():
...
 def wait_websocket(ws):
-    pod_name = session['pod_name']
-    svc_name = session['svc_name']
-    jupyter_token = session['jupyter_token']
-    log.info(f'received wait websocket for {svc_name} {pod_name}')
-    # wait for instance ready
+    notebook = session['notebook']
+
+    if notebook is None:
+        return
```
Comment (danking) [inline notebook2/notebook/notebook.py]: If the user opens a web socket without a notebook in the session, that's an error, no? We should let the web socket 500 and print a stack trace to the notebook logs.

File: notebook2/notebook/notebook.py (hunk 1 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Cotton and I had spoken about it: https://github.com/hail-is/hail/pull/5215#issuecomment-459225923 , it can be used with some restrictions. However, I think it may be more difficult than is outlined, because we need to take care of note only padding (`=`), but also `+` and `/`, which are in the output grammar (there some encodings that use `/` and `.`). Kubernetes allows `.`, `_`, and `-`, but requires those be in-between [a-Z0-9] characters, which I'm not sure we can guarantee.

In playing with character encodings, I can't actually get a + generated except in between alphanumeric ascii characters, but I would need to formally show that it isn't possible.

File: notebook2/notebook/notebook.py [hunk 9]
```diff
@@ -200,62 +335,62 @@ def root():
...
 def auth(requested_svc_name):
-    approved_svc_name = session.get('svc_name')
-    if approved_svc_name and approved_svc_name == requested_svc_name:
+    notebook = session.get('notebook')
+
+    if notebook is not None and notebook['svc_name'] == requested_svc_name:
         return '', 200
-    return '', 403
+
+    return '', 404
```
Comment (danking) [inline notebook2/notebook/notebook.py]: since we always return 403 or 200 from this endpoint, there's no leaked bit of information and 403 seems more semantically correct for something called `auth` (though it ought to be called authorized in retrospect).

File: notebook2/notebook/notebook.py (hunk 6 above)
Comment (danking) [reply notebook2/notebook/notebook.py]: we should `assert len(notebooks) == 1` and comment the issue number

File: notebook2/notebook/notebook.py [hunk 10]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+def read_svc_status(svc_name: str):
+    try:
+        # TODO: inspect exception for non-404
+        _ = k8s.read_namespaced_service(svc_name, 'default')
+        return 'Running'
+    except Exception:
+        return 'Deleted'
+
+
+def read_containers_status(container_statuses):
```
Comment (danking) [inline notebook2/notebook/notebook.py]: this method seems unnecessary. It is almost exactly 
```python
return [x.to_dict() for x in container_statuses]
```

Man it is infuriating that k8s uses `None` instead of `[]` for empty lists. I think we can change line 289 this and completely remove this method.
```python
    container_statuses = pod.status.container_statuses
    # ...
            'container_status': container_statuses or [x.to_dict() for x in container_statuses]
```

File: notebook2/notebook/notebook.py [hunk 11]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+
+        if condition.last_transition_time > maxDate and condition.status == "True":
+            maxCond = condition
+            maxDate = condition.last_transition_time
+
+    return {
+        "message": maxCond.message,
+        "reason": maxCond.reason,
+        "status": maxCond.status,
+        "type": maxCond.type}
```
Comment (danking) [inline notebook2/notebook/notebook.py]: This method can be:
```
conds = [c.to_dict for c in conds if c.status == "True"]
if not conds:
    return None
latest = max(conds, key=lambda x: x.last_transition_time)
return latest.to_dict()
```

File: notebook2/notebook/notebook.py [hunk 12]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+
+
+def marshall_notebooks(pods = [], svc_status = None):
```
Comment (danking) [inline notebook2/notebook/notebook.py]: mutable default arguments in python are dangerous and confusing, we should avoid them

File: notebook2/notebook/notebook.py [hunk 13]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+        notebook['url'] = f"/instance/{notebook['svc_name']}/?token={notebook['jupyter_token']}"
+
+        if svc_status is not None:
+            notebook['svc_status'] = svc_status
+        else:
+            notebook['svc_status'] = read_svc_status(notebook['svc_name'])
+
+        notebooks.append(notebook)
+
+    return notebooks
```
Comment (danking) [inline notebook2/notebook/notebook.py]: Mild preference for:
```
def pod_to_ui_dict(pod):
    notebook = {
        'name': pod.metadata.labels['name'],
        ... }
    # ...
    return notebook

def marshall_notebooks(pods, svc_status = None)
    return [pod_to_ui_dict(pod) for pod in pods]
```

Comment (danking) [reply notebook2/notebook/notebook.py]: Also mild inclination to call this `notebooks_for_ui` or something that conveys its restructuring the data intended for the UI.

File: notebook2/notebook/notebook.py [hunk 14]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+    pods = k8s.list_namespaced_pod(
+        namespace='default',
+        label_selector=f"user_id={user_id}",
+        timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
+
+    if len(pods) == 0:
+        return None
+
+    # Kubernetes service will reflect the change immediately, pod will not
+    notebooks = list(filter(lambda n: n['svc_status'] != 'Deleted', marshall_notebooks(pods)))
```
Comment (danking) [inline notebook2/notebook/notebook.py]: We can also check the pod for the presence of the deletionTimestamp metadata:
```bash
# k get pod dk-test-544786cc69-6rw6h -o yaml >y1 && \
  k delete pod dk-test-544786cc69-6rw6h && \
  k get pod dk-test-544786cc69-6rw6h -o yaml >y2
# diff y1 y2
7a8,9
>   deletionGracePeriodSeconds: 30
>   deletionTimestamp: 2019-03-04T15:47:31Z
21c23
<   resourceVersion: "32634684"
---
>   resourceVersion: "32634891"
# head y2
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/limit-ranger: 'LimitRanger plugin set: cpu request for container
      dk-test'
  creationTimestamp: 2019-03-04T15:46:10Z
  deletionGracePeriodSeconds: 30
  deletionTimestamp: 2019-03-04T15:47:31Z
```

**Comment by danking** [comment]
ah fuck. I accidentally approved instead of requesting changes and the CI moved way to fast for me.

**Comment by akotlar** [comment]
> ah fuck. I accidentally approved instead of requesting changes and the CI moved way to fast for me.

Tell me what you want done, and I'll submit a new PR

**Comment by akotlar** [comment]
Ah I see your comments, will address

**Comment by danking** [comment]
The relevant comments should all be listed here in the review

File: notebook2/notebook/notebook.py (hunk 14 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Useful, thanks!

File: notebook2/notebook/notebook.py (hunk 10 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Sounds good

File: notebook2/notebook/templates/notebook-state.html [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+                <b>{{"Waiting for " ~ notebook.condition.type}}</b>
+            {% else %}
+                <b>{{notebook.condition.type}}</b>
+            {% endif %}
+        </b>
+        </span>
+        <span className="small">
+        Pod: <b>{{notebook.pod_status}}</b>
+        </span>
+        <span className="small">
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/templates/notebook-state.html]: Shoot. Thanks

File: notebook2/notebook/notebook.py (hunk 10 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: So the method exists because I want to return only a part of the full response. A public interface for the web app, regardless of the representation returned by K1.

File: notebook2/notebook/notebook.py (hunk 11 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: So this function is slightly different, in two ways than the one your wrote. First, in the case that status is "False", and is the only status, I return it. This allows us to state we're "Waiting for: " + condition in the ui. Second, there are more values returned by V1PodCondition than I return here. And, a minor note, in my version, we iterate only 1 list, and create 0 lists. 

Given this, do you still prefer yours?

File: notebook2/notebook/notebook.py (hunk 12 above)
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Done

File: notebook2/notebook/notebook.py [hunk 16]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+            notebook['svc_status'] = svc_status
+        else:
+            notebook['svc_status'] = read_svc_status(notebook['svc_name'])
+
+        notebooks.append(notebook)
+
+    return notebooks
+
+
+def get_live_user_notebooks(user_id = None):
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: I generally prefer to return null/nil/None when a resource doesn't exist

File: notebook2/notebook/notebook.py [hunk 17]
```diff
@@ -188,6 +198,131 @@ def external_url_for(path):
...
+    if user_id is None:
+        return None
+
+    pods = k8s.list_namespaced_pod(
+        namespace='default',
+        label_selector=f"user_id={user_id}",
+        timeout_seconds=KUBERNETES_TIMEOUT_IN_SECONDS).items
```
[the remark this answers is not in the document]
Comment (akotlar) [reply notebook2/notebook/notebook.py]: Ugh, that is very confusing. I read it as "timeout for the list" operation, which is what this is...

File: notebook2/notebook/notebook.py (hunk 11 above)
Comment (danking) [reply notebook2/notebook/notebook.py]: Ah. I think that logic is more clearly described by:
```
if conds is None:
    return conds
return max(conds, key=lambda x: (x.last_transition_time, x.status)).to_dict()
```

File: notebook2/notebook/notebook.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (danking) [reply notebook2/notebook/notebook.py]: Using `None` instead of `[]` adds unnecessary condition checks when we're doing loop comprehensions.

Using both `None` and `[]` means we have two redundant ways to represent emptiness.

I don't see a compelling reason for using `None` instead of `[]` or both unless one of them has a distinct meaning.
